File: packages/mcp/src/credentials.ts
// 共享凭据文件读取。
//
// 与 Python SDK `sciverse.credentials` 同源契约：
// - 路径：`~/.sciverse/credentials.json`
// - 文件内容：{ token: string; endpoint?: string; saved_at?: string }
// - 解析顺序：[显式参数 → 环境变量 → 凭据文件 → 默认值]
//
// 写入由 Python CLI `sciverse auth login` 负责；MCP server / TS SDK
// 只读不写（让"装一次 Python CLI 后所有客户端形态都免传 token"链路成立）。
import { readFileSync, existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export interface StoredCredentials {
  token?: string;
  endpoint?: string;
  saved_at?: string;
}

export const DEFAULT_ENDPOINT = "https://api.sciverse.space";

/** 返回用户 home 目录。优先读环境变量 HOME/USERPROFILE 以便测试 override。 */
function getHomeDir(): string {
  return process.env.HOME ?? process.env.USERPROFILE ?? homedir();
}

export function credentialsPath(): string {
  return join(getHomeDir(), ".sciverse", "credentials.json");
}

/** 读凭据文件。文件不存在 / 解析失败 / 不是 object 时返回 null（不抛错）。 */
export function loadStoredCredentials(): StoredCredentials | null {
  const path = credentialsPath();
  if (!existsSync(path)) return null;
  try {
    const raw = readFileSync(path, "utf8");
    const data = JSON.parse(raw);
    if (data && typeof data === "object" && !Array.isArray(data)) {
      return data as StoredCredentials;
    }
    return null;
  } catch {
    return null;
  }
}

/** [显式 → env → 凭据文件] 顺序解析 token。都没有返回 null。 */
export function resolveToken(
  explicit?: string,
  env: NodeJS.ProcessEnv = process.env,
): string | null {
  if (explicit) return explicit;
  if (env.SCIVERSE_API_TOKEN) return env.SCIVERSE_API_TOKEN;
  const creds = loadStoredCredentials();
  if (creds?.token) return creds.token;
  return null;
}

/** [显式 → env → 凭据文件 → 默认值] 顺序解析 endpoint。 */
export function resolveEndpoint(
  explicit?: string,
  env: NodeJS.ProcessEnv = process.env,
): string {
  if (explicit) return explicit;
  if (env.SCIVERSE_BASE_URL) return env.SCIVERSE_BASE_URL;
  const creds = loadStoredCredentials();
  if (creds?.endpoint) return creds.endpoint;
  return DEFAULT_ENDPOINT;
}

File: packages/python/src/sciverse/credentials.py
"""共享凭据文件读写。

约定（与 MCP server / TS SDK 一致）：
- 路径：`~/.sciverse/credentials.json`
- 文件权限：0600（仅当前用户可读写）
- 内容：
  {
    "token": "sv-xxx",
    "endpoint": "https://api.sciverse.space",
    "saved_at": "2026-05-14T15:30:00Z"
  }

读取顺序（client.py / cli.py 共用）：
  1. 显式构造 client 传 token
  2. 环境变量 SCIVERSE_API_TOKEN
  3. 凭据文件
"""
from __future__ import annotations

import json
import os
import stat
from datetime import datetime, timezone
from pathlib import Path
from typing import TypedDict


class Credentials(TypedDict, total=False):
    token: str
    endpoint: str
    saved_at: str


DEFAULT_ENDPOINT = "https://api.sciverse.space"


def credentials_path() -> Path:
    return Path.home() / ".sciverse" / "credentials.json"


def load_credentials() -> Credentials | None:
    """读凭据文件。文件不存在 / 解析失败 / 不是 dict 时返回 None（不抛错）。"""
    path = credentials_path()
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(data, dict):
        return None
    return data  # type: ignore[return-value]


def save_credentials(token: str, endpoint: str = DEFAULT_ENDPOINT) -> Path:
    """保存凭据到 ~/.sciverse/credentials.json，文件权限设为 0600。返回 path。"""
    path = credentials_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    payload: Credentials = {
        "token": token,
        "endpoint": endpoint,
        "saved_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    # 0600
    try:
        path.chmod(stat.S_IRUSR | stat.S_IWUSR)
    except OSError:
        # Windows / 某些 FS 不支持 chmod；保留文件，不抛错
        pass
    return path


def delete_credentials() -> bool:
    """删凭据文件。文件不存在时返回 False，删除成功返回 True。"""
    path = credentials_path()
    if not path.exists():
        return False
    path.unlink()
    return True


def resolve_token(explicit: str | None = None) -> str | None:
    """按 [显式参数 → SCIVERSE_API_TOKEN → 凭据文件] 顺序返回 token，
    都没有时返回 None（不抛错；让调用方决定怎么提示用户）。"""
    if explicit:
        return explicit
    env_token = os.environ.get("SCIVERSE_API_TOKEN")
    if env_token:
        return env_token
    creds = load_credentials()
    if creds and creds.get("token"):
        return creds["token"]
    return None


def resolve_endpoint(explicit: str | None = None) -> str:
    """按 [显式参数 → SCIVERSE_BASE_URL → 凭据文件 → 默认值] 顺序返回 endpoint。"""
    if explicit:
        return explicit
    env_url = os.environ.get("SCIVERSE_BASE_URL")
    if env_url:
        return env_url
    creds = load_credentials()
    if creds and creds.get("endpoint"):
        return creds["endpoint"]
    return DEFAULT_ENDPOINT
